sway/commands/output.c

```diff
@@ -0,0 +1,251 @@
+				error = cmd_results_new(CMD_INVALID, "output", "Missing background file or color specification.");
+				goto fail;
+			}
+			if (i + 1 >= argc) {
+				error = cmd_results_new(CMD_INVALID, "output", "Missing background scaling mode or `solid_color`.");
+				goto fail;
+			}
+			if (strcasecmp(argv[i + 1], "solid_color") == 0) {
+				output->background = strdup(argv[argc - 2]);
+				output->background_option = strdup("solid_color");
+			} else {
+				// argv[i+j]=bg_option
+				bool valid = false;
+				char *mode;
+				size_t j;
+				for (j = 0; j < (size_t) (argc - i); ++j) {
+					mode = argv[i + j];
+					for (size_t k = 0; k < sizeof(bg_options) / sizeof(char *); ++k) {
+						if (strcasecmp(mode, bg_options[k]) == 0) {
+							valid = true;
+							break;
+						}
+					}
+					if (valid) {
+						break;
+					}
+				}
+				if (!valid) {
+					error = cmd_results_new(CMD_INVALID, "output", "Missing background scaling mode.");
+					goto fail;
+				}
+
+				char *src = join_args(argv + i, j);
+				if (wordexp(src, &p, 0) != 0 || p.we_wordv[0] == NULL) {
+					error = cmd_results_new(CMD_INVALID, "output", "Invalid syntax (%s)", src);
+					goto fail;
+				}
+				free(src);
+				src = p.we_wordv[0];
+				if (config->reading && *src != '/') {
+					char *conf = strdup(config->current_config);
+					if (conf) {
+						char *conf_path = dirname(conf);
+						src = malloc(strlen(conf_path) + strlen(src) + 2);
+						if (src) {
+							sprintf(src, "%s/%s", conf_path, p.we_wordv[0]);
+						} else {
+							sway_log(L_ERROR, "Unable to allocate background source");
+						}
+						free(conf);
+					} else {
+						sway_log(L_ERROR, "Unable to allocate background source");
+					}
+				}
+				if (!src || access(src, F_OK) == -1) {
+					error = cmd_results_new(CMD_INVALID, "output", "Background file unreadable (%s)", src);
+					wordfree(&p);
+					goto fail;
+				}
+
+				output->background = strdup(src);
+				output->background_option = strdup(mode);
+				if (src != p.we_wordv[0]) {
+					free(src);
+				}
+				wordfree(&p);
+
+				i += j;
+			}
+		}
+	}
+
+	i = list_seq_find(config->output_configs, output_name_cmp, name);
+	if (i >= 0) {
+		// merge existing config
+		struct output_config *oc = config->output_configs->items[i];
+		merge_output_config(oc, output);
+		free_output_config(output);
+		output = oc;
+	} else {
+		list_add(config->output_configs, output);
+	}
+
+	sway_log(L_DEBUG, "Config stored for output %s (enabled:%d) (%d x %d @ "
+			"%d, %d scale %d transform %d refresh_rate %f) (bg %s %s)",
+			output->name, output->enabled, output->width,
+			output->height, output->x, output->y, output->scale,
+			output->transform, output->refresh_rate,
+			output->background, output->background_option);
+
+	if (output->name) {
+		// Try to find the output container and apply configuration now. If
+		// this is during startup then there will be no container and config
+		// will be applied during normal "new output" event from wlc.
```

Comment:
event from what?

Reply:
Don't just copy+paste more than a few lines at a time, @emersion - make sure you read everything you import.